nau_openedx_extensions/coursecertificate/tests/test_celery.py

```diff
@@ -0,0 +1,415 @@
+
+    @patch(f"{COMMAND_MODULE_PATH}.process_service_certificates")
+    def test_handle_async_partial_failure(self, mock_task):
+        """Test async mode when some dispatches fail."""
+        # Setup - first call succeeds, second fails
+        mock_task.delay.side_effect = [
+            Mock(id="task-success"),
+            ValueError("Invalid config")
+        ]
+        config_yaml = yaml.dump(self.sample_config)
+
+        with patch("builtins.open", mock_open(read_data=config_yaml)):
+            self.command.handle_async({"config": "/fake/path/config.yml"})
+
+        # Verify
+        output = self.command.stdout.getvalue()
+        self.assertIn("Successfully dispatched: 1/2 tasks", output)
+        self.assertIn("Failed to dispatch: 1 tasks", output)
+        self.assertIn("task-success", output)
+
+    def test_handle_async_invalid_service_name(self):
+        """Test async mode with invalid service name."""
+        # Setup
+        config_yaml = yaml.dump(self.sample_config)
+
+        with patch("builtins.open", mock_open(read_data=config_yaml)):
+            with self.assertRaises(CommandError) as context:
+                self.command.handle_async({
+                    "config": "/fake/path/config.yml",
+                    "service_name": "nonexistent_service"
+                })
+
+        self.assertIn("Service 'nonexistent_service' not found", str(context.exception))
+
+    @patch(f"{COMMAND_MODULE_PATH}.process_service_certificates")
+    def test_dispatch_service_task_safely_success(self, mock_task):
+        """Test successful task dispatch."""
+        # Setup
+        mock_task.delay.return_value = Mock(id="task-789")
+        service_config = self.sample_config["NAU_SEND_COURSE_CERTIFICATE_CONFIG"][0]
+
+        # Execute
+        result = self.command.dispatch_service_task_safely(service_config, {"dry_run": False})
+
+        # Verify
+        self.assertIsNotNone(result)
+        self.assertEqual(result["service_name"], "test_service")
+        self.assertEqual(result["task_id"], "task-789")
+
+        output = self.command.stdout.getvalue()
+        self.assertIn("Task dispatched for service 'test_service'", output)
+
+    @patch(f"{COMMAND_MODULE_PATH}.process_service_certificates")
+    def test_dispatch_service_task_safely_connection_error(self, mock_task):
+        """Test task dispatch with connection error."""
+        # Setup
+        mock_task.delay.side_effect = ConnectionError("Broker connection failed")
+        service_config = self.sample_config["NAU_SEND_COURSE_CERTIFICATE_CONFIG"][0]
+
+        # Execute
+        result = self.command.dispatch_service_task_safely(service_config, {"dry_run": False})
+
+        # Verify
+        self.assertIsNone(result)
+        output = self.command.stdout.getvalue()
+        self.assertIn("Failed to dispatch task for service 'test_service'", output)
+        self.assertIn("Broker connection failed", output)
```

Comment:
Normally, to avoid leaving these comments (`# Setup`, `# Execute`, `# Verify`), we simply separate the code using a line break, for example:

```suggestion
    @patch(f"{COMMAND_MODULE_PATH}.process_service_certificates")
    def test_dispatch_service_task_safely_connection_error(self, mock_task):
        """Test task dispatch with connection error."""
        mock_task.delay.side_effect = ConnectionError("Broker connection failed")
        service_config = self.sample_config["NAU_SEND_COURSE_CERTIFICATE_CONFIG"][0]

        result = self.command.dispatch_service_task_safely(service_config, {"dry_run": False})

        self.assertIsNone(result)
        output = self.command.stdout.getvalue()
        self.assertIn("Failed to dispatch task for service 'test_service'", output)
        self.assertIn("Broker connection failed", output)
```

Reply:
Done